File: backend/app/models/paiement.py
from datetime import datetime

from sqlalchemy import Boolean, Column, DateTime, Float, ForeignKey, Integer, String, Text
from sqlalchemy.orm import relationship

from app.core.database import Base


class Paiement(Base):
    __tablename__ = "paiements"

    id = Column(
        Integer,
        primary_key=True,
        index=True,
    )

    facture_id = Column(
        Integer,
        ForeignKey("factures.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )

    montant = Column(
        Float,
        nullable=False,
    )

    date_paiement = Column(
        DateTime,
        nullable=False,
        default=datetime.utcnow,
    )

    mode_paiement = Column(
        String(50),
        nullable=False,
    )

    operateur = Column(
        String(30),
        nullable=True,
    )

    transaction_id = Column(
        String(100),
        nullable=True,
        unique=True,
        index=True,
    )

    numero_client = Column(
        String(30),
        nullable=True,
    )

    taux_commission = Column(
        Float,
        nullable=False,
        default=2.0,
    )

    montant_commission = Column(
        Float,
        nullable=False,
        default=0.0,
    )

    montant_total = Column(
        Float,
        nullable=False,
        default=0.0,
    )

    frais = Column(
        Float,
        nullable=False,
        default=0,
    )

    reference = Column(
        String(100),
        nullable=True,
        index=True,
    )

    statut = Column(
        String(30),
        nullable=False,
        default="En attente",
    )

    notes = Column(
        Text,
        nullable=True,
    )

    actif = Column(
        Boolean,
        nullable=False,
        default=True,
    )

    created_at = Column(
        DateTime,
        nullable=False,
        default=datetime.utcnow,
    )

    updated_at = Column(
        DateTime,
        nullable=False,
        default=datetime.utcnow,
        onupdate=datetime.utcnow,
    )

    facture = relationship(
        "Facture",
        back_populates="paiements",
    )

File: frontend/lib/models/paiement.dart
class Paiement {
  final int id;
  final int factureId;
  final double montant;
  final String modePaiement;
  final String? operateur;
  final String? transactionId;
  final String? numeroClient;
  final double frais;
  final double tauxCommission;
  final double montantCommission;
  final double montantTotal;
  final String? reference;
  final DateTime? datePaiement;
  final String statut;
  final String? notes;
  final bool actif;
  final DateTime createdAt;
  final DateTime updatedAt;

  const Paiement({
    required this.id,
    required this.factureId,
    required this.montant,
    required this.modePaiement,
    this.operateur,
    this.transactionId,
    this.numeroClient,
    required this.frais,
    required this.tauxCommission,
    required this.montantCommission,
    required this.montantTotal,
    this.reference,
    this.datePaiement,
    required this.statut,
    this.notes,
    required this.actif,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Paiement.fromJson(Map<String, dynamic> json) {
    return Paiement(
      id: json['id'] as int,
      factureId: json['facture_id'] as int,
      montant: (json['montant'] as num).toDouble(),
      modePaiement: json['mode_paiement'] as String,
      operateur: json['operateur'] as String?,
      transactionId: json['transaction_id'] as String?,
      numeroClient: json['numero_client'] as String?,
      frais: (json['frais'] as num?)?.toDouble() ?? 0,
      tauxCommission:
          (json['taux_commission'] as num?)?.toDouble() ?? 2,
      montantCommission:
          (json['montant_commission'] as num?)?.toDouble() ?? 0,
      montantTotal:
          (json['montant_total'] as num?)?.toDouble() ?? 0,
      reference: json['reference'] as String?,
      datePaiement: json['date_paiement'] != null
          ? DateTime.parse(json['date_paiement'] as String)
          : null,
      statut: json['statut'] as String? ?? 'En attente',
      notes: json['notes'] as String?,
      actif: json['actif'] as bool? ?? true,
      createdAt: DateTime.parse(json['created_at'] as String),
      updatedAt: DateTime.parse(json['updated_at'] as String),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'facture_id': factureId,
      'montant': montant,
      'mode_paiement': modePaiement,
      'operateur': operateur,
      'transaction_id': transactionId,
      'numero_client': numeroClient,
      'frais': frais,
      'taux_commission': tauxCommission,
      'montant_commission': montantCommission,
      'montant_total': montantTotal,
      'reference': reference,
      'date_paiement': datePaiement?.toIso8601String(),
      'statut': statut,
      'notes': notes,
      'actif': actif,
      'created_at': createdAt.toIso8601String(),
      'updated_at': updatedAt.toIso8601String(),
    };
  }
}
